Can't build tests on win-ifort. Fails on linker with strptime

Hello,

I have been using your library with no problems in unix with gfortran, but time has come to compile things in Windows using ifort.

Compiling the library runs, although some warnings do show up. Compiling the tests fails with 

Severity	Code	Description	Project	File	Line	Suppression State
Error		error LNK2019: unresolved external symbol strptime referenced in function MOD_DATETIME_mp_STRPTIME		datetime.lib(mod_datetime.obj)

I'm a bit stuck, any ideas?

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -16,9 +16,17 @@ Fortran_MODULE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/include
 ARCHIVE_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib
 LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib)
 
+if(WIN32)
+  enable_language(CXX)
+  add_library(strptime OBJECT strptime.cpp)
+  set_target_properties(strptime PROPERTIES CXX_STANDARD 11 CXX_STANDARD_REQUIRED true)
+  target_link_libraries(datetime PRIVATE strptime)
+endif()
+
 # tests
 add_executable(datetime_tests src/tests/datetime_tests.f90)
 target_link_libraries(datetime_tests datetime)
 set_target_properties(datetime_tests PROPERTIES
-RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
+RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin
+LINKER_LANGUAGE Fortran)
 add_test(NAME datetime_tests COMMAND $<TARGET_FILE:datetime_tests>)

--- a/src/lib/datetime.f90
+++ b/src/lib/datetime.f90
@@ -2,7 +2,7 @@
 ! datetime-fortran - A Fortran library for date and time manipulation
 ! Copyright (c) 2013-2017, Wavebit Scientific LLC
 ! All rights reserved.
-! 
+!
 ! Licensed under the BSD-3 clause license. See LICENSE for details.
 !
 module datetime_module

--- a/src/lib/mod_datetime.f90
+++ b/src/lib/mod_datetime.f90
@@ -85,7 +85,7 @@ type :: datetime
   procedure,pass(d0),private :: datetime_plus_timedelta
   procedure,pass(d0),private :: timedelta_plus_datetime
   procedure,pass(d0),private :: datetime_minus_datetime
-  procedure,pass(d0),private :: datetime_minus_timedelta 
+  procedure,pass(d0),private :: datetime_minus_timedelta
   procedure,pass(d0),private :: eq
   procedure,pass(d0),private :: neq
   procedure,pass(d0),private :: gt
@@ -123,14 +123,14 @@ pure elemental type(datetime) function datetime_constructor(year,month,&
   integer,          intent(in),optional :: year        !! year
   integer,          intent(in),optional :: month       !! month
   integer,          intent(in),optional :: day         !! day
-  integer,          intent(in),optional :: hour        !! hour 
+  integer,          intent(in),optional :: hour        !! hour
   integer,          intent(in),optional :: minute      !! minute
   integer,          intent(in),optional :: second      !! second
   integer,          intent(in),optional :: millisecond !! millisecond
   real(kind=real64),intent(in),optional :: tz          !! timezone offset in hours
 
   if(present(year))then
-    datetime_constructor % year = year    
+    datetime_constructor % year = year
   else
     datetime_constructor % year = 1
   endif
@@ -181,7 +181,7 @@ endfunction datetime_constructor
 
 
 
-! datetime getters  
+! datetime getters
 !=======================================================================
 
 pure elemental integer function getYear(self)
@@ -273,12 +273,12 @@ pure elemental subroutine addMilliseconds(self,ms)
 endsubroutine addMilliseconds
 
 
-! datetime-bound methods 
+! datetime-bound methods
 !=======================================================================
 
 pure elemental subroutine addSeconds(self,s)
 
-  !! Adds an integer number of seconds to self. Called by `datetime` 
+  !! Adds an integer number of seconds to self. Called by `datetime`
   !! addition (`+`) and subtraction (`-`) operators.
 
   class(datetime),intent(inout) :: self !! `datetime` instance
@@ -304,7 +304,7 @@ endsubroutine addSeconds
 
 pure elemental subroutine addMinutes(self,m)
 
-  !! Adds an integer number of minutes to self. Called by `datetime` 
+  !! Adds an integer number of minutes to self. Called by `datetime`
   !! addition (`+`) and subtraction (`-`) operators.
 
   class(datetime),intent(inout) :: self !! `datetime` instance
@@ -330,7 +330,7 @@ endsubroutine addMinutes
 
 pure elemental subroutine addHours(self,h)
 
-  !! Adds an integer number of hours to self. Called by `datetime` 
+  !! Adds an integer number of hours to self. Called by `datetime`
   !! addition (`+`) and subtraction (`-`) operators.
 
   class(datetime),intent(inout) :: self !! `datetime` instance
@@ -356,7 +356,7 @@ endsubroutine addHours
 
 pure elemental subroutine addDays(self,d)
 
-  !! Adds an integer number of dayss to self. Called by `datetime` 
+  !! Adds an integer number of dayss to self. Called by `datetime`
   !! addition (`+`) and subtraction (`-`) operators.
 
   class(datetime),intent(inout) :: self !! `datetime` instance
@@ -383,7 +383,7 @@ pure elemental subroutine addDays(self,d)
       self % day = self % day+daysInMonth(self % month,self % year)
     else
       exit
-    endif 
+    endif
   enddo
 
 endsubroutine addDays
@@ -395,7 +395,7 @@ pure elemental character(len=23) function isoformat(self,sep)
   !! Returns character string with time in ISO 8601 format.
 
   class(datetime), intent(in)          :: self !! `datetime instance`
-  character(len=1),intent(in),optional :: sep  
+  character(len=1),intent(in),optional :: sep
     !! separator character, 'T' is default
 
   character(len=1) :: separator
@@ -406,8 +406,8 @@ pure elemental character(len=23) function isoformat(self,sep)
     separator = 'T'
   endif
 
-  ! TODO below is a bit cumbersome and was implemented 
-  ! at a time before the interface to strftime. Now we 
+  ! TODO below is a bit cumbersome and was implemented
+  ! at a time before the interface to strftime. Now we
   ! could do something like:
   !
   ! isoformat = self % strftime('%Y-%m-%d'//separator//'%H:%M:%S')
@@ -426,8 +426,8 @@ endfunction isoformat
 
 pure elemental logical function isValid(self)
 
-  !! Checks whether the `datetime` instance has valid component values. 
-  !! Returns `.true.` if the `datetime` instance is valid, and `.false.` 
+  !! Checks whether the `datetime` instance has valid component values.
+  !! Returns `.true.` if the `datetime` instance is valid, and `.false.`
   !! otherwise.
 
   class(datetime),intent(in) :: self !! `datetime` instance
@@ -450,23 +450,23 @@ pure elemental logical function isValid(self)
     isValid = .false.
     return
   endif
- 
-  if(self % hour < 0 .or. self % hour > 23)then    
+
+  if(self % hour < 0 .or. self % hour > 23)then
     isValid = .false.
     return
   endif
 
-  if(self % minute < 0 .or. self % minute > 59)then    
+  if(self % minute < 0 .or. self % minute > 59)then
     isValid = .false.
     return
   endif
 
-  if(self % second < 0 .or. self % second > 59)then    
+  if(self % second < 0 .or. self % second > 59)then
     isValid = .false.
     return
   endif
 
-  if(self % millisecond < 0 .or. self % millisecond > 999)then    
+  if(self % millisecond < 0 .or. self % millisecond > 999)then
     isValid = .false.
     return
   endif
@@ -477,7 +477,7 @@ endfunction isValid
 
 type(datetime) function now()
 
-  !! Returns a `datetime` instance with current time. 
+  !! Returns a `datetime` instance with current time.
   !! No input arguments.
 
   character(len=5)     :: zone
@@ -507,7 +507,7 @@ endfunction now
 
 pure elemental integer function weekday(self)
 
-  !! Returns the day of the week calculated using Zeller's congruence. 
+  !! Returns the day of the week calculated using Zeller's congruence.
   !! Returned value is an integer scalar in the range [0-6], such that:
   !!
   !! 0: Sunday
@@ -558,7 +558,7 @@ pure elemental integer function isoweekday(self)
   class(datetime),intent(in) :: self !! `datetime` instance
 
   isoweekday = self % weekday()
-  
+
   if (isoweekday == 0) then
     isoweekday = 7
   end if
@@ -632,8 +632,8 @@ endfunction isoweekdayShort
 
 function isocalendar(self)
 
-  !! Returns an array of 3 integers, year, week number, and week day, 
-  !! as defined by ISO 8601 week date. Essentially a wrapper around C 
+  !! Returns an array of 3 integers, year, week number, and week day,
+  !! as defined by ISO 8601 week date. Essentially a wrapper around C
   !! `strftime` function.
 
   class(datetime),intent(in) :: self !! `datetime` instance
@@ -644,7 +644,7 @@ function isocalendar(self)
   character(len=20)    :: string
 
   rc = c_strftime(string,len(string),'%G %V %u'//c_null_char,&
-                  self % tm())  
+                  self % tm())
 
   read(unit=string(1:4),fmt='(I4)')year
   read(unit=string(6:7),fmt='(I2)')week
@@ -658,10 +658,10 @@ endfunction isocalendar
 
 integer function secondsSinceEpoch(self)
 
-  !! Returns an integer number of seconds since the UNIX Epoch, 
-  !! `1970-01-01 00:00:00`. Note that this is a wrapper around C's 
-  !! `strftime('%s')`, so the number of seconds will reflect the time 
-  !! zone of the local machine on which the function is being called. 
+  !! Returns an integer number of seconds since the UNIX Epoch,
+  !! `1970-01-01 00:00:00`. Note that this is a wrapper around C's
+  !! `strftime('%s')`, so the number of seconds will reflect the time
+  !! zone of the local machine on which the function is being called.
 
   class(datetime),intent(in) :: self !! `datetime` instance
 
@@ -748,7 +748,7 @@ endfunction tzOffset
 
 pure elemental type(datetime) function utc(self)
 
-  !! Returns the `datetime` instance at Coordinated Universal Time (UTC). 
+  !! Returns the `datetime` instance at Coordinated Universal Time (UTC).
 
   class(datetime),intent(in) :: self !! `datetime` instance
 
@@ -890,8 +890,8 @@ endfunction datetime_minus_datetime
 
 pure elemental logical function gt(d0,d1)
 
-  !! `datetime` comparison operator that eturns `.true.` if `d0` is 
-  !! greater than `d1` and `.false.` otherwise. Overloads the 
+  !! `datetime` comparison operator that eturns `.true.` if `d0` is
+  !! greater than `d1` and `.false.` otherwise. Overloads the
   !! operator `>`.
 
   class(datetime),intent(in) :: d0 !! lhs `datetime` instance
@@ -965,7 +965,7 @@ endfunction gt
 
 pure elemental logical function lt(d0,d1)
 
-  !! `datetime` comparison operator that returns `.true.` if `d0` is 
+  !! `datetime` comparison operator that returns `.true.` if `d0` is
   !! less than `d1` and `.false.` otherwise. Overloads the operator `<`.
 
   class(datetime),intent(in) :: d0 !! lhs `datetime` instance
@@ -979,7 +979,7 @@ endfunction lt
 
 pure elemental logical function eq(d0,d1)
 
-  !! `datetime` comparison operator that returns `.true.` if `d0` is 
+  !! `datetime` comparison operator that returns `.true.` if `d0` is
   !! equal to `d1` and `.false.` otherwise. Overloads the operator `==`.
 
   class(datetime),intent(in) :: d0 !! lhs `datetime` instance
@@ -1005,7 +1005,7 @@ endfunction eq
 
 pure elemental logical function neq(d0,d1)
 
-  !! `datetime` comparison operator that eturns `.true.` if `d0` is 
+  !! `datetime` comparison operator that eturns `.true.` if `d0` is
   !! not equal to `d1` and `.false.` otherwise. Overloads the operator `/=`.
 
   class(datetime),intent(in) :: d0 !! lhs `datetime` instance
@@ -1019,8 +1019,8 @@ endfunction neq
 
 pure elemental logical function ge(d0,d1)
 
-  !! `datetime` comparison operator. Returns `.true.` if `d0` is greater 
-  !! than or equal to `d1` and `.false.` otherwise. Overloads the 
+  !! `datetime` comparison operator. Returns `.true.` if `d0` is greater
+  !! than or equal to `d1` and `.false.` otherwise. Overloads the
   !! operator `>=`.
 
   class(datetime),intent(in) :: d0 !! lhs `datetime` instance
@@ -1034,8 +1034,8 @@ endfunction ge
 
 pure elemental logical function le(d0,d1)
 
-  !! `datetime` comparison operator. Returns `.true.` if `d0` is less 
-  !! than or equal to `d1`, and `.false.` otherwise. Overloads the 
+  !! `datetime` comparison operator. Returns `.true.` if `d0` is less
+  !! than or equal to `d1`, and `.false.` otherwise. Overloads the
   !! operator `<=`.
 
   class(datetime),intent(in) :: d0 !! lhs `datetime` instance
@@ -1065,9 +1065,9 @@ endfunction isLeapYear
 
 pure function datetimeRange(d0,d1,t)
 
-  !! Given start and end `datetime` instances `d0` and `d1` and time 
-  !! increment as `timedelta` instance `t`, returns an array of 
-  !! `datetime` instances. The number of elements is the number of whole 
+  !! Given start and end `datetime` instances `d0` and `d1` and time
+  !! increment as `timedelta` instance `t`, returns an array of
+  !! `datetime` instances. The number of elements is the number of whole
   !! time increments contained between datetimes `d0` and `d1`.
 
   type(datetime), intent(in) :: d0 !! start time
@@ -1113,8 +1113,8 @@ pure elemental integer function daysInMonth(month,year)
 
   if(month < 1 .or. month > 12)then
     ! Should raise an error and abort here, however we want to keep
-    ! the pure and elemental attributes. Make sure this function is 
-    ! called with the month argument in range. 
+    ! the pure and elemental attributes. Make sure this function is
+    ! called with the month argument in range.
     daysInMonth = 0
     return
   endif
@@ -1147,7 +1147,7 @@ endfunction daysInYear
 
 pure elemental real(kind=real64) function date2num(d)
 
-  !! Given a datetime instance d, returns number of days since 
+  !! Given a datetime instance d, returns number of days since
   !! `0001-01-01 00:00:00`, taking into account the timezone offset.
 
   type(datetime),intent(in) :: d !! `datetime` instance
@@ -1174,17 +1174,17 @@ pure elemental real(kind=real64) function date2num(d)
            + d_utc % hour*h2d  &
            + d_utc % minute*m2d&
            + (d_utc % second+1e-3_real64*d_utc % millisecond)*s2d
- 
+
 endfunction date2num
 
 
 
 pure elemental type(datetime) function num2date(num)
 
-  !! Given number of days since `0001-01-01 00:00:00`, returns a 
+  !! Given number of days since `0001-01-01 00:00:00`, returns a
   !! correspoding `datetime` instance.
 
-  real(kind=real64),intent(in) :: num 
+  real(kind=real64),intent(in) :: num
     !! number of days since `0001-01-01 00:00:00`
 
   integer :: year,month,day,hour,minute,second,millisecond
@@ -1247,7 +1247,7 @@ endfunction num2date
 type(datetime) function strptime(str,format)
 
   !! A wrapper function around C/C++ strptime function.
-  !! Returns a `datetime` instance. 
+  !! Returns a `datetime` instance.
 
   character(len=*),intent(in) :: str    !! time string
   character(len=*),intent(in) :: format !! time format
@@ -1264,7 +1264,7 @@ endfunction strptime
 
 pure elemental type(datetime) function tm2date(ctime)
 
-  !! Given a `tm_struct` instance, returns a corresponding `datetime` 
+  !! Given a `tm_struct` instance, returns a corresponding `datetime`
   !! instance.
 
   type(tm_struct),intent(in) :: ctime !! C-style time struct

--- a/src/tests/datetime_tests.f90
+++ b/src/tests/datetime_tests.f90
@@ -5,7 +5,7 @@ MODULE datetime_tests
 !
 !=======================================================================
 USE datetime_module
-USE iso_fortran_env,ONLY:REAL64
+USE iso_fortran_env,ONLY:REAL64, stdout=>output_unit, stderr=>error_unit
 
 IMPLICIT NONE
 
@@ -13,9 +13,6 @@ PRIVATE
 
 PUBLIC :: assert
 PUBLIC :: test_datetime
-
-INTEGER,PARAMETER :: STDOUT = 6
-INTEGER,PARAMETER :: STDERR = 0
 
 CONTAINS
 
@@ -74,7 +71,7 @@ ENDSUBROUTINE initialize_tests
 SUBROUTINE report_tests(tests,test_failed)
 !=======================================================================
 !
-! Takes the test counter as input and reports the total number of test 
+! Takes the test counter as input and reports the total number of test
 ! passes and failures.
 !
 !=======================================================================
@@ -88,7 +85,7 @@ SUBROUTINE report_tests(tests,test_failed)
 
   nsuccess = 0
   nfailure = 0
-  DO n = 1,ntests  
+  DO n = 1,ntests
     IF(tests(n))THEN
       nsuccess = nsuccess + 1
     ELSE
@@ -136,7 +133,7 @@ SUBROUTINE test_datetime
 
   WRITE(UNIT=STDOUT,FMT='(A)')
 
-  ! Test counter; 
+  ! Test counter;
   ! modify if adding new tests
   ntests = 191
 
@@ -177,7 +174,7 @@ SUBROUTINE test_datetime
                  == datetime(2013,12,31,23,59,59,0),                  &
                  'datetime + timedelta(seconds = -1)')
   n = n+1
-                   
+
   ! Increment minutes
   tests(n) = assert(datetime(2014,1,1,0,0,0,0) + timedelta(minutes=1)&
                  == datetime(2014,1,1,0,1,0,0),                      &
@@ -195,7 +192,7 @@ SUBROUTINE test_datetime
                  == datetime(2014,1,1,1,0,0,0),                    &
                  'datetime + timedelta(hours = 1)')
   n = n+1
-             
+
   ! Decrement hours
   tests(n) = assert(datetime(2014,1,1,0,0,0,0) + timedelta(hours=-1)&
                  == datetime(2013,12,31,23,0,0,0),                  &
@@ -258,62 +255,62 @@ SUBROUTINE test_datetime
   ! Test subtracting into previous months:
   tests(n) = assert(datetime(2014,2, 1,0,0,0,0) + timedelta(days=-1)&
                  == datetime(2014,1,31,0,0,0,0),                    &
-                 'decrement datetime into January')  
+                 'decrement datetime into January')
   n = n+1
 
   tests(n) = assert(datetime(2014,3, 1,0,0,0,0) + timedelta(days=-1)&
                  == datetime(2014,2,28,0,0,0,0),                    &
-                 'decrement datetime into February')  
+                 'decrement datetime into February')
   n = n+1
-  
+
   tests(n) = assert(datetime(2014,4, 1,0,0,0,0) + timedelta(days=-1)&
                  == datetime(2014,3,31,0,0,0,0),                    &
-                 'decrement datetime into March')  
+                 'decrement datetime into March')
   n = n+1
-  
+
   tests(n) = assert(datetime(2014,5, 1,0,0,0,0) + timedelta(days=-1)&
                  == datetime(2014,4,30,0,0,0,0),                    &
-                 'decrement datetime into April')  
+                 'decrement datetime into April')
   n = n+1
-  
+
   tests(n) = assert(datetime(2014,6, 1,0,0,0,0) + timedelta(days=-1)&
                  == datetime(2014,5,31,0,0,0,0),                    &
-                 'decrement datetime into May')  
+                 'decrement datetime into May')
   n = n+1
-  
+
   tests(n) = assert(datetime(2014,7, 1,0,0,0,0) + timedelta(days=-1)&
                  == datetime(2014,6,30,0,0,0,0),                    &
-                 'decrement datetime into June')  
+                 'decrement datetime into June')
   n = n+1
-  
+
   tests(n) = assert(datetime(2014,8, 1,0,0,0,0) + timedelta(days=-1)&
                  == datetime(2014,7,31,0,0,0,0),                    &
-                 'decrement datetime into July')  
+                 'decrement datetime into July')
   n = n+1
-  
+
   tests(n) = assert(datetime(2014,9, 1,0,0,0,0) + timedelta(days=-1)&
                  == datetime(2014,8,31,0,0,0,0),                    &
-                 'decrement datetime into August')  
+                 'decrement datetime into August')
   n = n+1
-  
+
   tests(n) = assert(datetime(2014,10, 1,0,0,0,0) + timedelta(days=-1)&
                  == datetime(2014, 9,30,0,0,0,0),                    &
-                 'decrement datetime into September')  
+                 'decrement datetime into September')
   n = n+1
-  
+
   tests(n) = assert(datetime(2014,11, 1,0,0,0,0) + timedelta(days=-1)&
                  == datetime(2014,10,31,0,0,0,0),                    &
                  'decrement datetime into October')
   n = n+1
-  
+
   tests(n) = assert(datetime(2014,12, 1,0,0,0,0) + timedelta(days=-1)&
                  == datetime(2014,11,30,0,0,0,0),                    &
-                 'decrement datetime into November')  
+                 'decrement datetime into November')
   n = n+1
 
   tests(n) = assert(datetime(2015, 1, 1,0,0,0,0) + timedelta(days=-1)&
                  == datetime(2014,12,31,0,0,0,0),                    &
-                 'decrement datetime into December')  
+                 'decrement datetime into December')
   n = n+1
   !---------------------------------------------------------------------
 
@@ -381,7 +378,7 @@ SUBROUTINE test_datetime
                  >= datetime(2014,1,2,3,4,5,6),&
                  'datetime >= datetime (greater)')
   n = n+1
-   
+
   tests(n) = assert(datetime(2014,1,2,3,4,5,6) &
                  >= datetime(2014,1,2,3,4,5,6),&
                  'datetime >= datetime (equal)')
@@ -396,7 +393,7 @@ SUBROUTINE test_datetime
                  <= datetime(2014,1,2,4,4,5,6),&
                  'datetime <= datetime (less)')
   n = n+1
-   
+
   tests(n) = assert(datetime(2014,1,2,3,4,5,6) &
                  <= datetime(2014,1,2,3,4,5,6),&
                  'datetime <= datetime (equal)')
@@ -456,121 +453,121 @@ SUBROUTINE test_datetime
   a = datetime(2005,1,1)
   tests(n) = assert(ALL(a % isocalendar() == [2004,53,6]),&
                    'datetime(2005,1,1) % isocalendar() == [2004,53,6]')
-  n = n+1 
+  n = n+1
 
   a = datetime(2005,1,2)
   tests(n) = assert(ALL(a % isocalendar() == [2004,53,7]),&
                    'datetime(2005,1,2) % isocalendar() == [2004,53,7]')
-  n = n+1 
+  n = n+1
 
   a = datetime(2005,12,31)
   tests(n) = assert(ALL(a % isocalendar() == [2005,52,6]),&
                    'datetime(2005,12,31) % isocalendar() == [2005,52,6]')
-  n = n+1 
+  n = n+1
 
   a = datetime(2007,1,1)
   tests(n) = assert(ALL(a % isocalendar() == [2007,1,1]),&
                    'datetime(2007,1,1) % isocalendar() == [2007,1,1]')
-  n = n+1 
+  n = n+1
 
   a = datetime(2007,12,30)
   tests(n) = assert(ALL(a % isocalendar() == [2007,52,7]),&
                    'datetime(2007,12,30) % isocalendar() == [2007,52,7]')
-  n = n+1 
+  n = n+1
 
   a = datetime(2007,12,31)
   tests(n) = assert(ALL(a % isocalendar() == [2008,1,1]),&
                    'datetime(2007,12,31) % isocalendar() == [2008,1,1]')
-  n = n+1 
+  n = n+1
 
   a = datetime(2008,1,1)
   tests(n) = assert(ALL(a % isocalendar() == [2008,1,2]),&
                    'datetime(2008,1,1) % isocalendar() == [2008,1,2]')
-  n = n+1 
+  n = n+1
 
   a = datetime(2008,12,28)
   tests(n) = assert(ALL(a % isocalendar() == [2008,52,7]),&
                    'datetime(2008,12,28) % isocalendar() == [2008,52,7]')
-  n = n+1 
+  n = n+1
 
   a = datetime(2008,12,29)
   tests(n) = assert(ALL(a % isocalendar() == [2009,1,1]),&
                    'datetime(2008,12,29) % isocalendar() == [2009,1,1]')
-  n = n+1 
+  n = n+1
 
   a = datetime(2008,12,30)
   tests(n) = assert(ALL(a % isocalendar() == [2009,1,2]),&
                    'datetime(2008,12,30) % isocalendar() == [2009,1,2]')
-  n = n+1 
+  n = n+1
 
   a = datetime(2008,12,31)
   tests(n) = assert(ALL(a % isocalendar() == [2009,1,3]),&
                    'datetime(2008,12,31) % isocalendar() == [2009,1,3]')
-  n = n+1 
+  n = n+1
 
   a = datetime(2009,1,1)
   tests(n) = assert(ALL(a % isocalendar() == [2009,1,4]),&
                    'datetime(2009,1,1) % isocalendar() == [2009,1,4]')
-  n = n+1 
+  n = n+1
 
   a = datetime(2009,12,31)
   tests(n) = assert(ALL(a % isocalendar() == [2009,53,4]),&
                    'datetime(2009,12,31) % isocalendar() == [2009,53,4]')
-  n = n+1 
+  n = n+1
 
   a = datetime(2010,1,1)
   tests(n) = assert(ALL(a % isocalendar() == [2009,53,5]),&
                    'datetime(2010,1,1) % isocalendar() == [2009,53,5]')
-  n = n+1 
+  n = n+1
 
   a = datetime(2010,1,2)
   tests(n) = assert(ALL(a % isocalendar() == [2009,53,6]),&
                    'datetime(2010,1,2) % isocalendar() == [2009,53,6]')
-  n = n+1 
+  n = n+1
 
   a = datetime(2010,1,3)
   tests(n) = assert(ALL(a % isocalendar() == [2009,53,7]),&
                    'datetime(2010,1,3) % isocalendar() == [2009,53,7]')
-  n = n+1 
+  n = n+1
   !---------------------------------------------------------------------
 
   WRITE(UNIT=STDOUT,FMT='(71("-"))')
 
   !---------------------------------------------------------------------
   ! datetime % isValid()
-  
+
   a = datetime(1,1,1)
   tests(n) = assert(a % isValid(),'datetime(1,1,1) is valid')
   n = n+1
-  
+
   a = datetime(0,1,1)
   tests(n) = assert(.NOT. a % isValid(),'datetime(0,1,1) is not valid')
   n = n+1
-  
+
   a = datetime(-1,1,1)
   tests(n) = assert(.NOT. a % isValid(),'datetime(-1,1,1) is not valid')
   n = n+1
-  
+
   a = datetime(2014,1,1)
   tests(n) = assert(a % isValid(),'datetime(2014,1,1) is valid')
   n = n+1
-  
+
   a = datetime(2014,0,1)
   tests(n) = assert(.NOT. a % isValid(),'datetime(2014,0,1) is not valid')
   n = n+1
-  
+
   a = datetime(2014,1,0)
   tests(n) = assert(.NOT. a % isValid(),'datetime(2014,1,0) is not valid')
   n = n+1
-  
+
   a = datetime(2014,2,28)
   tests(n) = assert(a % isValid(),'datetime(2014,2,28) is valid')
   n = n+1
-  
+
   a = datetime(2014,2,29)
   tests(n) = assert(.NOT. a % isValid(),'datetime(2014,2,29) is not valid')
   n = n+1
-  
+
   a = datetime(2012,2,29)
   tests(n) = assert(a % isValid(),'datetime(2012,2,29) is valid')
   n = n+1
@@ -607,7 +604,7 @@ SUBROUTINE test_datetime
   WRITE(UNIT=STDOUT,FMT='(71("-"))')
 
   !---------------------------------------------------------------------
-  ! datetime % secondsSinceEpoch 
+  ! datetime % secondsSinceEpoch
   a = datetime(1970,1,1,0,0,0)
 
   ! First get local machine offset in seconds
@@ -628,7 +625,7 @@ SUBROUTINE test_datetime
   n = n+1
 
   !---------------------------------------------------------------------
-  
+
   WRITE(UNIT=STDOUT,FMT='(71("-"))')
 
   !---------------------------------------------------------------------
@@ -1033,7 +1030,7 @@ SUBROUTINE test_datetime
   n = n+1
 
   !---------------------------------------------------------------------
-    
+
   WRITE(UNIT=STDOUT,FMT='(71("-"))')
 
   !---------------------------------------------------------------------
@@ -1127,7 +1124,7 @@ SUBROUTINE test_datetime
   !---------------------------------------------------------------------
 
   WRITE(UNIT=STDOUT,FMT='(71("-"))')
-  
+
   !---------------------------------------------------------------------
   ! daysInYear
 
@@ -1146,7 +1143,7 @@ SUBROUTINE test_datetime
   !---------------------------------------------------------------------
 
   WRITE(UNIT=STDOUT,FMT='(71("-"))')
-  
+
   !---------------------------------------------------------------------
   ! daysInMonth
 
@@ -1192,7 +1189,7 @@ SUBROUTINE test_datetime
   !---------------------------------------------------------------------
 
   WRITE(UNIT=STDOUT,FMT='(71("-"))')
-  
+
   !---------------------------------------------------------------------
   ! clock tests
 
@@ -1204,15 +1201,15 @@ SUBROUTINE test_datetime
 
   tests(n) = assert(.NOT. c % started,'clock % started == F (before tick)')
   n = n+1
-  
+
   CALL c % tick()
 
   tests(n) = assert(c % started,'clock % started == T (after 1 tick)')
   n = n+1
-  
+
   tests(n) = assert(.NOT. c % stopped,'clock % stopped == F (after 1 tick)')
   n = n+1
-  
+
   ! Tick 23 times
   DO i = 1,23
     CALL c % tick()
@@ -1230,14 +1227,14 @@ SUBROUTINE test_datetime
 
   tests(n) = assert(.NOT. c % started,'clock % started == F (after reset)')
   n = n+1
-  
+
   tests(n) = assert(.NOT. c % started,'clock % stopped == F (after reset)')
   n = n+1
-  
+
   tests(n) = assert(c % currentTime == c % startTime,&
                     'clock % currentTime == clock % startTime (after reset)')
   n = n+1
-  
+
   !---------------------------------------------------------------------
   WRITE(UNIT=STDOUT,FMT='(71("-"))')
 
@@ -1262,7 +1259,7 @@ PROGRAM run_tests
 ! Unit test driver for datetime-fortran.
 !
 !=======================================================================
-USE datetime_tests 
+USE datetime_tests
 IMPLICIT NONE
 
 CALL test_datetime()

--- a/strptime.cpp
+++ b/strptime.cpp
@@ -1,0 +1,22 @@
+// https://stackoverflow.com/a/33542189
+#include <time.h>
+#include <iomanip>
+#include <sstream>
+
+extern "C" char* strptime(const char* s,
+                          const char* f,
+                          struct tm* tm) {
+  // Isn't the C++ standard lib nice? std::get_time is defined such that its
+  // format parameters are the exact same as strptime. Of course, we have to
+  // create a string stream first, and imbue it with the current C locale, and
+  // we also have to make sure we return the right things if it fails, or
+  // if it succeeds, but this is still far simpler an implementation than any
+  // of the versions in any of the C standard libraries.
+  std::istringstream input(s);
+  input.imbue(std::locale(setlocale(LC_ALL, nullptr)));
+  input >> std::get_time(tm, f);
+  if (input.fail()) {
+    return nullptr;
+  }
+  return (char*)(s + input.tellg());
+}

